src/CloudNative.CloudEvents.Protobuf/ProtobufEventFormatter.cs

```diff
@@ -0,0 +1,412 @@
+
+        /// <inheritdoc />
+        public override ReadOnlyMemory<byte> EncodeBinaryModeEventData(CloudEvent cloudEvent)
+        {
+            Validation.CheckCloudEventArgument(cloudEvent, nameof(cloudEvent));
+
+            if (cloudEvent.Data is null)
+            {
+                return Array.Empty<byte>();
+            }
+            if (cloudEvent.DataContentType is string dataContentType && dataContentType.StartsWith("text/") && cloudEvent.Data is string text)
+            {
+                ContentType contentType = new ContentType(dataContentType);
+                return MimeUtilities.GetEncoding(contentType).GetBytes(text);
+            }
+            if (cloudEvent.Data is byte[] bytes)
+            {
+                return bytes;
+            }
+            throw new ArgumentException($"{nameof(ProtobufEventFormatter)} cannot serialize data of type {cloudEvent.Data.GetType()} with content type '{cloudEvent.DataContentType}'");
+        }
+
+        /// <inheritdoc />
+        public override ReadOnlyMemory<byte> EncodeStructuredModeMessage(CloudEvent cloudEvent, out ContentType contentType)
+        {
+            var proto = ConvertToProto(cloudEvent, nameof(cloudEvent));
+            contentType = new ContentType(StructuredMediaType)
+            {
+                CharSet = Encoding.UTF8.WebName
+            };
+            return proto.ToByteArray();
+        }
+
+        /// <inheritdoc />
+        public override IReadOnlyList<CloudEvent> DecodeBatchModeMessage(Stream body, ContentType? contentType, IEnumerable<CloudEventAttribute>? extensionAttributes)
+        {
+            Validation.CheckNotNull(body, nameof(body));
+            var batchProto = CloudEventBatch.Parser.ParseFrom(body);
+            return batchProto.Events.Select(proto => ConvertFromProto(proto, extensionAttributes, nameof(body))).ToList();
+        }
+
+        /// <inheritdoc />
+        public override Task<IReadOnlyList<CloudEvent>> DecodeBatchModeMessageAsync(Stream body, ContentType? contentType, IEnumerable<CloudEventAttribute>? extensionAttributes)
+        {
+            return base.DecodeBatchModeMessageAsync(body, contentType, extensionAttributes);
+        }
+
+        /// <inheritdoc />
+        public override CloudEvent DecodeStructuredModeMessage(Stream messageBody, ContentType? contentType, IEnumerable<CloudEventAttribute>? extensionAttributes)
+        {
+            Validation.CheckNotNull(messageBody, nameof(messageBody));
+            return ConvertFromProto(V1.CloudEvent.Parser.ParseFrom(messageBody), extensionAttributes, nameof(messageBody));
+        }
+
+        /// <inheritdoc />
+        public override Task<CloudEvent> DecodeStructuredModeMessageAsync(Stream body, ContentType? contentType, IEnumerable<CloudEventAttribute>? extensionAttributes)
+        {
+            return base.DecodeStructuredModeMessageAsync(body, contentType, extensionAttributes);
```

Comment:
nit: await

Reply:
Actually in this case I'm just going to remove this override and the batch one earlier... we shouldn't have them.